hudi-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadata.java

```diff
@@ -919,4 +886,42 @@ private Map<String, HoodieRecord<HoodieMetadataPayload>> fetchBaseFileAllRecords
       return SecondaryIndexKeyUtils.getRecordKeyFromSecondaryIndexKey(record.getRecordKey());
     }, record -> record));
   }
+
+  @VisibleForTesting
+  public static Map<String, String> reverseLookupSecondaryKeysInternal(List<String> recordKeys,
+                                                                       Map<String, HoodieRecord<HoodieMetadataPayload>> baseFileRecords,
+                                                                       HoodieMetadataLogRecordReader logRecordScanner) {
+    Map<String, String> recordKeyMap = new HashMap<>();
+    Set<String> keySet = new TreeSet<>(recordKeys);
+    Set<String> deletedRecordsFromLogs = new HashSet<>();
+    Map<String, HoodieRecord<HoodieMetadataPayload>> logRecordsMap = new HashMap<>();
+    // Note that: we read the log records from the oldest to the latest!!!
+    // If we change the read order, we need update the following logic accordingly.
+    logRecordScanner.getRecords().forEach(record -> {
+      String recordKey = SecondaryIndexKeyUtils.getRecordKeyFromSecondaryIndexKey(record.getRecordKey());
+      HoodieMetadataPayload payload = record.getData();
+      if (!payload.isDeleted()) { // process only valid records.
+        if (keySet.contains(recordKey)) {
+          logRecordsMap.put(recordKey, record);
+        }
+      } else {
+        // When and Only when the latest log record is non-tombstone, logRecordMap contains its recordKey.
+        logRecordsMap.remove(recordKey);
+        deletedRecordsFromLogs.add(recordKey);
+      }
+    });
+
+    // Return non-tombstone records from the log files.
+    logRecordsMap.forEach((key, value) -> recordKeyMap.put(
+        key, SecondaryIndexKeyUtils.getSecondaryKeyFromSecondaryIndexKey(value.getRecordKey())));
+    // Return non-tombstone records from the base file.
+    if (baseFileRecords != null) {
+      baseFileRecords.forEach((key, value) -> {
+        if (!deletedRecordsFromLogs.contains(key)) {
+          recordKeyMap.put(key, SecondaryIndexKeyUtils.getSecondaryKeyFromSecondaryIndexKey(value.getRecordKey()));
```

Comment:
why we are not merging the records from base to log records?  I understand in case of secondary index, a record is either created or deleted and there is no real merge, but lets keep the original code as is and just add fixes on top of that.

Reply:
For eg, I went to master and made edits as below. all tests you have added in this patch worked. 

```
+  @VisibleForTesting
+  public static Map<String, String> reverseLookupSecondaryKeysInternal(List<String> keySet, Map<String, HoodieRecord<HoodieMetadataPayload>> baseFileRecords,
+                                                                HoodieMetadataLogRecordReader logRecordScanner) {
+    Set<String> deletedRecordsFromLogs = new HashSet<>();
+    // Map of recordKey (primaryKey) -> log record that is not deleted for all input recordKeys
+    Map<String, HoodieRecord<HoodieMetadataPayload>> logRecordsMap = new HashMap<>();
+    logRecordScanner.getRecords().forEach(record -> {
+      String recordKey = SecondaryIndexKeyUtils.getRecordKeyFromSecondaryIndexKey(record.getRecordKey());
+      HoodieMetadataPayload payload = record.getData();
+      if (!payload.isDeleted()) { // process only valid records.
+        if (keySet.contains(recordKey)) {
+          logRecordsMap.put(recordKey, record);
+          deletedRecordsFromLogs.remove(recordKey); // we can check if its present and then remove if need be
+        }
+      } else {
+        deletedRecordsFromLogs.add(recordKey);
+        logRecordsMap.remove(recordKey);
+      }
+    });
+
+    Map<String, String> recordKeyMap = new HashMap<>();
+    if (baseFileRecords == null || baseFileRecords.isEmpty()) {
+      logRecordsMap.forEach((key1, value1) -> {
+        if (!value1.getData().isDeleted() && !deletedRecordsFromLogs.contains(key1)) {
+          recordKeyMap.put(key1, SecondaryIndexKeyUtils.getSecondaryKeyFromSecondaryIndexKey(value1.getRecordKey()));
+        }
+      });
+    } else {
+      // Iterate over all provided log-records, merging them into existing records
+      logRecordsMap.forEach((key1, value1) -> baseFileRecords.merge(key1, value1, (oldRecord, newRecord) -> {
+        Option<HoodieRecord<HoodieMetadataPayload>> mergedRecord = HoodieMetadataPayload.combineSecondaryIndexRecord(oldRecord, newRecord);
+        return mergedRecord.orElse(null);
+      }));
+      baseFileRecords.forEach((key, value) -> {
+        if (!deletedRecordsFromLogs.contains(key)) {
+          recordKeyMap.put(key, SecondaryIndexKeyUtils.getSecondaryKeyFromSecondaryIndexKey(value.getRecordKey()));
+        }
+      });
+    }
     return recordKeyMap;
   }
```

Reply:
@nsivabalan , the it is the merge that cause the bug. Suppose we have `br` and a non-delete `lr`. In this case, after the merge, `lr` is returned. Now it will go through the ` if (!deletedRecordsFromLogs.contains(key))`, since `deletedRecordsFromLogs` contains its rk, `lr` is gone.

Reply:
Ok, will adopt the above code.

Reply:
@nsivabalan , the reason is that we discuss here is that the logic of this function is too complex to understand. so we should try to make it as simple as possible.

Reply:
Both are fine IMO, but maybe I am biased due to prior work. Ideally, I would like to retain the semantics of `combineSecondaryIndexRecord` but the current patch is also simple to understand. Plus, if we filter out tombstone records before calling `combineSecondaryIndexRecord` then there is no point in keeping it. Please remove this method.